contracts/wawaka/common/contract/attestation.cpp

```diff
@@ -0,0 +1,363 @@
+    if (! KeyValueStore::privileged_get("ContractCode.Nonce", code_nonce))
+        return rsp.error("failed to retreive privileged value for ContractCode.Nonce");
+
+    ww::value::Structure result(CONTRACT_CODE_METADATA_SCHEMA);
+    result.set_string("code_nonce", code_nonce.c_str());
+
+    return rsp.value(result, false);
+}
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+// NAME: add_endpoint
+//
+// Contract method to verify the attestation of a contract object
+// and add it to the set of "trusted endpoints"
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+bool ww::contract::attestation::add_endpoint(
+    const Message& msg,
+    const Environment& env,
+    Response& rsp)
+{
+    // to add an endpoint we do need the ledger key
+    std::string ledger_key;
+    if (! get_ledger_key(ledger_key) && ledger_key.length() > 0)
+        return rsp.error("contract has not been initialized");
+
+    // extract the parameters
+    ASSERT_SUCCESS(rsp, msg.validate_schema(ADD_ENDPOINT_PARAM_SCHEMA),
+                   "invalid request, missing required parameters");
+
+    const std::string contract_id(msg.get_string("contract_id"));
+    const std::string ledger_code_hash(msg.get_string("ledger_attestation.contract_code_hash"));
+    const std::string ledger_meta_hash(msg.get_string("ledger_attestation.metadata_hash"));
+    const std::string ledger_signature(msg.get_string("ledger_attestation.signature"));
+    const std::string verifying_key(msg.get_string("contract_metadata.verifying_key"));
+    const std::string encryption_key(msg.get_string("contract_metadata.encryption_key"));
+    const std::string code_nonce(msg.get_string("contract_code_metadata.code_nonce"));
+
+    // we are going to assume that the invoker of this method is the creator
+    // of the contract being added so the creator id will come from the environment
+    const std::string creator(env.originator_id_);
+
+    // verify the ledger's signature on the metadata_hash and code_hash
+    {
+        ww::types::ByteArray buffer;
+        std::copy(contract_id.begin(), contract_id.end(), std::back_inserter(buffer));
+        std::copy(creator.begin(), creator.end(), std::back_inserter(buffer));
+        std::copy(ledger_code_hash.begin(), ledger_code_hash.end(), std::back_inserter(buffer));
+        std::copy(ledger_meta_hash.begin(), ledger_meta_hash.end(), std::back_inserter(buffer));
+
+        ww::types::ByteArray signature;
+        if (! ww::crypto::b64_decode(ledger_signature, signature))
+            return rsp.error("failed to decode ledger signature");
+        if (! ww::crypto::ecdsa::verify_signature(buffer, ledger_key, signature))
+            return rsp.error("failed to verify ledger signature");
+    }
+
+    // verify that the code hash matches the code hash in the ledger
+    {
+        // we compute the merkle root using the hash of our own code
+        // so we know the hash of the other if it matches
```

Comment:
I found this comment to be a bit confusing. It's not immediately clear from this comment that `set_code_hash()` is expected be called before calling `add_endpoint()`. I think it would be helpful to clarify this assumption.